kasa/klapprotocol.py

```diff
@@ -0,0 +1,500 @@
+            server_hash.hex(),
+        )
+
+        local_seed_auth_hash = self._sha256(self._local_seed + self.local_auth_hash)
+
+        # Check the response from the device
+        if local_seed_auth_hash == server_hash:
+            _LOGGER.debug("handshake1 hashes match")
+            return remote_seed, self.local_auth_hash
+        else:
+            _LOGGER.debug(
+                "Expected %s got %s in handshake1.  Checking if blank auth is a match",
+                local_seed_auth_hash.hex(),
+                server_hash.hex(),
+            )
+
+            blank_auth = Credentials(username="", password="")
+            blank_auth_hash = self.generate_auth_hash(blank_auth)
+            blank_seed_auth_hash = self._sha256(self._local_seed + blank_auth_hash)
+            if blank_seed_auth_hash == server_hash:
+                _LOGGER.debug(
+                    "Server response doesn't match our expected hash on ip %s but an authentication with blank credentials matched",
+                    self.host,
+                )
+                return remote_seed, blank_auth_hash
+            else:
+                kasa_setup_auth = Credentials(
+                    username=self.KASA_SETUP_EMAIL, password=self.KASA_SETUP_PASSWORD
+                )
+                kasa_setup_auth_hash = self.generate_auth_hash(kasa_setup_auth)
+                kasa_setup_seed_auth_hash = self._sha256(
+                    self._local_seed + kasa_setup_auth_hash
+                )
+                if kasa_setup_seed_auth_hash == server_hash:
+                    self.local_auth_hash = kasa_setup_auth_hash
+                    _LOGGER.debug(
+                        "Server response doesn't match our expected hash on ip %s but an authentication with kasa setup credentials matched",
+                        self.host,
+                    )
+                    return remote_seed, kasa_setup_auth_hash
+                else:
+                    self.authentication_failed = True
+                    msg = f"Server response doesn't match our challenge on ip {self.host} {self.discovery_data}"
+                    _LOGGER.debug(msg)
+                    raise AuthenticationException(msg)
+
+    async def perform_handshake2(self, session, remote_seed, auth_hash) -> None:
+        """Perform handshake2.  Sets authentication_failed based on success/failure."""
+        # Handshake 2 has the following payload:
+        #    sha256(serverBytes | authenticator)
+
+        url = f"http://{self.host}/app/handshake2"
+
+        payload = self._sha256(remote_seed + auth_hash)
+
+        response_status, response_data = await self.session_post(
+            session, url, data=payload
+        )
+
+        cookie = self.jar.filter_cookies(url).get(self.TP_SESSION_COOKIE_NAME)
+        _LOGGER.debug(
+            f"Handshake2 posted {datetime.datetime.now()}.  Host is {self.host}, Session cookie is {cookie}, Response status is {response_status}, Request was {payload!r}"
```

Comment:
https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/logging-not-lazy.html

Reply:
All logging now lazy